cmd/expired-authz-purger/main.go

```diff
@@ -95,34 +96,50 @@ func (p *expiredAuthzPurger) purge(table string, yes bool, purgeBefore time.Time
 		}
 	}
 
+	sem := make(chan bool, parallelism)
 	for _, id := range ids {
-		// Delete challenges + authorization. We delete challenges first and fail out
-		// if that doesn't succeed so that we don't ever orphan challenges which would
-		// require a relatively expensive join to then find.
-		_, err := p.db.Exec("DELETE FROM challenges WHERE authorizationID = ?", id)
-		if err != nil {
-			return err
-		}
-		var query string
-		switch table {
-		case "pendingAuthorizations":
-			query = "DELETE FROM pendingAuthorizations WHERE id = ?"
-		case "authz":
-			query = "DELETE FROM authz WHERE id = ?"
-		}
-		_, err = p.db.Exec(query, id)
-		if err != nil {
-			return err
-		}
+		sem <- true
+		go func(id string) {
+			err := deleteAuthorization(p.db, table, id)
+			if err != nil {
+				p.log.AuditErr(fmt.Sprintf("Deleting %s: %s", id, err))
+			}
+			<-sem
+		}(id)
+	}
+	for i := 0; i < cap(sem); i++ {
```

Comment:
I'm probably missing something really straight-forward (forgive me, working on little sleep today!) but why is this follow-up loop to write true across the sem channel capacity necessary?

Reply:
Without it, the main for loop will finish while there are still goroutines outstanding. Each write to the channel will block until one of the goroutines finishes, so this second loop will complete only after all the goroutines are done.